File: LeetCode/palindrome_number.go
// // O(log(x)), O(log(x))
// func isPalindrome(x int) bool {
//     if x < 0 {
//         return false
//     } else if x < 10 {
//         return true
//     } else {
//         var xList []int
//         num := x
//         for num > 0 {
//             xList = append(xList, num % 10)
//             num /= 10
//         }
        
//         len_xlist := len(xList)
//         half_len_xlist := len_xlist / 2
//         for i := 0; i < half_len_xlist; i++ {
//             if xList[i] != xList[len_xlist - i - 1] {
//                 return false
//             }
//         }
        
//         return true
//     }
// }

// O(log(x)), O(1)
func isPalindrome(x int) bool {
    if x < 0 {
        return false
    } else if x < 10 {
        return true
    } else {
        tmp_x := x
        reverse := 0
        for tmp_x > 0 {
            reverse = reverse * 10 + tmp_x % 10
            tmp_x /= 10
        }
        
        return x == reverse
    }
}

File: LeetCode/palindrome_number.py
class Solution:
#     # O(log(x)), O(log(x))
#     def isPalindrome(self, x: int) -> bool:
#         if x < 0: return False
#         if x < 10: return True
#         if x % 10 == 0: return False
        
#         x_list = []
#         while x > 0:
#             val = x % 10
#             x //= 10
#             x_list.append(val)
        
#         mid = -1
#         len_x = len(x_list)
#         stack = list()
#         mid = len_x // 2
#         for i, n in enumerate(x_list):
#             if i < mid: stack.append(n)
#             else:
#                 if i == mid and len_x % 2 != 0: continue
#                 else:
#                     if stack[-1] == n: stack.pop()
#         return len(stack) == 0

    # O(log(x)), O(1)
    def isPalindrome(self, x: int) -> bool:
        if x < 0: return False
        if x < 10: return True
        if x % 10 == 0: return False
        
        tmp_x = x
        reverse = 0
        while tmp_x > 0:
            reverse = reverse * 10 + tmp_x % 10
            tmp_x //= 10
        return x == reverse
        